support for H740P Enhanced HBA mode in dracclient

mpardhi23 opened the pull request:
Hey Chris,
This PR is having Implementation of code for getting raid controller mode and setting required raid controller mode in python dracclient for adding support for H740P Enhanced HBA mode.
Tested this code for H70P controller and also updated unit test cases.

cdearborn commented on dracclient/client.py:
```diff
@@ -753,6 +753,55 @@ def list_raid_controllers(self):
         """
         return self._raid_mgmt.list_raid_controllers()
 
+    def list_raid_cntrl_settings(self, by_name=False):
```
Would you mind changing the above method name to:
list_raid_controller_settings()?

mpardhi23 replied:
done Chris.

cdearborn commented on dracclient/client.py:
```diff
@@ -753,6 +753,55 @@ def list_raid_controllers(self):
+        """
+        return self._raid_mgmt.list_raid_cntrl_settings(by_name)
+
+    def set_raid_cntrl_settings(self, settings, raid_fqdd):
```
Would you mind changing the above method name to:
set_raid_controller_settings()?

mpardhi23 replied:
sure

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,220 @@
+                 read_only, fqdd, possible_values):
+        """Creates RAIDEnumerableAttribute object
+
+        :param name: name of the RAID attribute
```
Can you add a :param for instance_id here?

mpardhi23 replied:
Done

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,220 @@
+        instance_id = utils.get_wsman_resource_attr(
+            raid_attr_xml, namespace, 'InstanceID')
+        current_value = utils.get_wsman_resource_attr(
+            raid_attr_xml, namespace, 'CurrentValue', nullable=True)
```
I think there is an issue here.  I was looking at DCIM_RAIDEnumeration, and some of the attributes can have multiple CurrentValue's, like RAIDSupportedRAIDLevels.  I believe the code above will just take the first one found and set that as the current_value.  I think we're going to need to change current_value to current_values and make it an array.

cdearborn replied:
Wow, this is a much bigger problem.  Both CurrentValue and PendingValue are listed as an array in the 3.30.30.30 MOFs but we treat them as single values everywhere.

mpardhi23 replied:
Yes Agree Chris. I have checked some of the attributes in RAIDEnumeration , which is having multiple current value. I have updated code for that.

mpardhi23 replied:
> Wow, this is a much bigger problem. Both CurrentValue and PendingValue are listed as an array in the 3.30.30.30 MOFs but we treat them as single values everywhere.

I have checked the RAIDEnumeration attributes, we are having multiple current values and possible values but I have not found multiple pending values for any attribute. But need to test it once while setting the new raid setting when attribute value needed multiple current values to be set. Please correct me If I am getting wrong here.

mpardhi23 replied:
> I think there is an issue here. I was looking at DCIM_RAIDEnumeration, and some of the attributes can have multiple CurrentValue's, like RAIDSupportedRAIDLevels. I believe the code above will just take the first one found and set that as the current_value. I think we're going to need to change current_value to current_values and make it an array.

Yes Agree Chris. I have checked some of the attributes in RAIDEnumeration , which is having multiple current value. I have updated code for that. Doing one round of testing with new changes.

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,220 @@
+
+        return utils.list_settings(self.client, self.NAMESPACES, by_name)
+
+    def set_raid_cntrl_settings(self, new_settings, raid_fqdd):
```
Can you change "cntrl" to "controller" in the method name above?

mpardhi23 replied:
Done

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,220 @@
         """
         self.client = client
 
+    def list_raid_cntrl_settings(self, by_name=False):
```
Can you change "cntrl" to "controller" in the method name above?

mpardhi23 replied:
Done

mpardhi23 commented:
Hey Chris,
I have updated PR with the below changes as suggested by you and also tested those changes.
    1. List RAID settings should return an array for the value of current_value.
    2. Set RAID settings should accept only a scalar value like other resources case.

cdearborn commented on dracclient/client.py:
```diff
@@ -753,6 +753,57 @@ def list_raid_controllers(self):
+        :param settings: a dictionary containing the proposed values, with
+                         each key being the name of attribute and the value
+                         being the proposed value.
+        :param raid_fqdd: the FQDD of the RAID.
```
can you change the above to: the FQDD of the RAID controller. ?

cdearborn commented on dracclient/client.py:
```diff
@@ -753,6 +753,57 @@ def list_raid_controllers(self):
+    def get_raid_controller_mode(self):
+        """Returns the current RAID controller mode
+
+        :returns: the current RAIDController mode
```
can you change the above to "RAID controller"?

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,229 @@
+    def get_raid_controller_mode(self):
+        """Returns the current RAID controller mode
+
+        :returns: the current RAIDController mode
```
can you change the above from "RAIDController" to "RAID controller"?

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,229 @@
+                                                      pending_value,
+                                                      read_only, fqdd)
+
+        self.current_value = current_value
```
Can you move current_value into RAIDAttribute and move the parsing for it there as well?  Please keep the parameters for __init__ in the same order that we have them for BIOSAttribute, etc.

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,229 @@
+        super(RAIDStringAttribute, self).__init__(name, instance_id,
+                                                  pending_value,
+                                                  read_only, fqdd)
+        self.current_value = current_value
```
Can you move current_value into RAIDAttribute and move the parsing for it there as well? Please keep the parameters for init in the same order that we have them for BIOSAttribute, etc.

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,229 @@
+        return cls(raid_attr.name, raid_attr.instance_id,
+                   raid_attr.pending_value, raid_attr.read_only,
+                   raid_attr.fqdd, current_value, min_length, max_length)
+
```
Can you add a RAIDIntegerAttribute class and include this in the list() and set() methods?

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,229 @@
 class RAIDManagement(object):
 
+    NAMESPACES = [(uris.DCIM_RAIDEnumeration, RAIDEnumerableAttribute),
+                  (uris.DCIM_RAIDString, RAIDStringAttribute)]
```
Can you add a RAIDIntegerAttribute class?

cdearborn commented on dracclient/resources/uris.py:
```diff
@@ -97,6 +97,12 @@
+
+DCIM_RAIDString = ('http://schemas.dell.com/wbem/wscim/1/cim-schema/2/'
+                   'DCIM_RAIDString')
+
```
Can you add a DCIM_RAIDInteger URI?

cdearborn commented on dracclient/tests/utils.py:
```diff
@@ -247,6 +247,12 @@ def load_wsman_xml(name):
+        'ok': load_wsman_xml('raid_enumeration-enum-ok')
+    },
+    uris.DCIM_RAIDString: {
+        'ok': load_wsman_xml('raid_string-enum-ok')
```
Can you add an entry for DCIM_RAIDInteger here?

cdearborn commented on dracclient/tests/test_raid.py:
```diff
@@ -112,6 +113,216 @@ def setUp(self):
+                       'wait_until_idrac_is_ready', spec_set=True,
+                       autospec=True)
+    def test_set_raid_controller_settings_with_unchanged_attr(
+            self, mock_requests, mock_wait_until_idrac_is_ready):
```
This test case is really confusing.  Looking in the sample xml file, there is a RAIDdefaultWritePolicy for every virtual disk.  This makes me think that setting attributes by attribute name must be invalid, because how would the idrac know which attribute to change?  Does it change them all?  Does it error out?

cdearborn replied:
To avoid this confusion, it might be better to pick an attribute that occurs only once, like RAIDControllerBootMode or something like that.

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,229 @@
+        be created.
+        :param new_settings: a dictionary containing the proposed values, with
+                         each key being the name of attribute and the value
+                         being the proposed value.
```
I'm wondering if we have a significant bug here.  Looking at DCIM_RAIDEnumeration, there are a lot of attributes that have the same AttributeName, but different InstanceIDs.  In utils.set_settings, it calls list_settings(by_name=True).  This will result in a much smaller list of attributes than if it had been called with by_name=False, which will make everything based off current_settings behave strangely.
If a user passed in RAIDdefaultWritePolicy for an attribute name, what would the iDRAC do?  I would think it would have to error out since it couldn't determine what attribute to set.  What if the user wanted to set that attribute on 2 different virtual disks?  How would they do that?  Would they pass in the InstanceID instead of the name?  If they did that, then wouldn't set_settings fail because it wouldn't find that attribute?  Is it valid to put the InstanceID in the SetAttributes packet instead of the attribute name?

mpardhi23 replied:
Yes right, for many of the attributes AttributeName is same ,so iDRAC will not be able to find the corresponding attribute which need to be set.

I had tried set_settings using AttributeName 'RAIDdefaultWritePolicy' which is having different InstanceIDs with list_settings(by_name=True) with new_settings like settings = {'RAIDdefaultWritePolicy':'WriteBack'} which results in below error:

  File "/usr/lib/python2.7/site-packages/dracclient/utils.py", line 441, in set_settings
    wait_for_idrac=wait_for_idrac)
  File "/usr/lib/python2.7/site-packages/dracclient/client.py", line 1333, in invoke
    raise exceptions.DRACOperationFailed(drac_messages=messages)
dracclient.exceptions.DRACOperationFailed: DRAC operation failed. Messages: ['Invalid Attribute Name RAIDdefaultWritePolicy']

I had also tried set_settings using instanceID  with list_settings(by_name=False) as well , it also getting failed with below error:

(Pdb) utils.find_xml(resp, 'MessageID', resource_uri).text
'STOR040'
  File "/usr/lib/python2.7/site-packages/dracclient/utils.py", line 441, in set_settings
    wait_for_idrac=wait_for_idrac)
  File "/usr/lib/python2.7/site-packages/dracclient/client.py", line 1333, in invoke
    raise exceptions.DRACOperationFailed(drac_messages=messages)
dracclient.exceptions.DRACOperationFailed: DRAC operation failed. Messages: ['Invalid Attribute Name RAID.Slot.6-1:RAIDRequestedControllerMode']

mpardhi23 commented:
Thanks Chris for your review comments! I have moved current_value  in init and also added the RAIDIntegerAttribute class in raid.py as suggested locally and modified the code for setting RAID settings. And also tested these changes.

mahendraKamble commented:
Hi Chris,
I have updated PR with Monica's following changes.
1. Added RAIDIntegerAttribute class for listing RAIDInteger Attributes.
2. Moved the current_value into RAIDAttribute and the parsing for it in the same order as BIOS, NIC attributes.

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,295 @@
+                                  raid_fqdd,
+                                  by_name=False)
+
+    def get_raid_controller_mode(self):
```
We should pass in raid_fqdd here in case there is more than 1 RAID controller in the server

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,295 @@
+        :raises: DRACOperationFailed on error reported back by the DRAC
+                 interface
+        """
+        raid_cntrl_attr = self.list_raid_controller_settings(by_name=True)
```
We should remove the by_name parameter here so that the attributes are listed by instance ID to eliminate any possibility of confusion should there be multiple RAID controllers.

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,295 @@
+        """
+        raid_cntrl_attr = self.list_raid_controller_settings(by_name=True)
+
+        raid_cntrl_mode = raid_cntrl_attr.get('RAIDCurrentControllerMode')
```
The above line should get the attribute by instance ID using the passed in FQDD and hard coded "RAIDCurrentControllerMode"

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,295 @@
         """
         self.client = client
 
+    def list_raid_controller_settings(self, by_name=False):
```
I don't think a "by_name" parameter here makes sense because there are several attributes that are per virtual disk, which means there could be multiple attributes with the same name, but different instance IDs for a single RAID controller.  Let's remove this parameter to keep users from shooting themselves in the foot.

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,295 @@
+        """
+
+        return utils.list_settings(self.client, self.NAMESPACES,
+                                   by_name=by_name)
```
We should pass by_name=False on the line above

cdearborn commented on dracclient/resources/raid.py:
```diff
@@ -98,15 +98,295 @@
+        """List the RAID configuration settings
+
+        :param by_name: Controls whether returned dictionary uses RAID
+                        attribute name or instance_id as key.
```
We should remove the above by_name param

cdearborn commented on dracclient/client.py:
```diff
@@ -753,6 +753,57 @@ def list_raid_controllers(self):
         """
         return self._raid_mgmt.list_raid_controllers()
 
+    def list_raid_controller_settings(self, by_name=False):
```
I don't think a "by_name" parameter here makes sense because there are several attributes that are per virtual disk, which means there could be multiple attributes with the same name, but different instance IDs for a single RAID controller.  Let's remove this parameter to keep users from shooting themselves in the foot.

cdearborn commented on dracclient/client.py:
```diff
@@ -753,6 +753,57 @@ def list_raid_controllers(self):
+        """List the RAID configuration settings
+
+        :param by_name: Controls whether returned dictionary uses RAID
+                        attribute name or instance_id as key.
```
Let's remove the above by_name param

cdearborn commented on dracclient/client.py:
```diff
@@ -753,6 +753,57 @@ def list_raid_controllers(self):
+        :raises: DRACOperationFailed on error reported back by the DRAC
+                 interface
+        """
+        return self._raid_mgmt.list_raid_controller_settings(by_name)
```
Lets remove the above by_name parameter

cdearborn commented on dracclient/utils.py:
```diff
@@ -396,21 +423,26 @@ def set_settings(settings_type,
 
-    return build_return_dict(doc, resource_uri)
+    return build_return_dict(doc, resource_uri,
+                             include_commit_required=include_commit_required)
```
We should never include commit_required in the results.  This parameter was deprecated and removed from python-dracclient.

cdearborn commented on dracclient/utils.py:
```diff
@@ -132,7 +132,9 @@ def get_all_wsman_resource_attrs(doc, resource_uri, attr_name, nullable=False):
-                      is_reboot_required_value=None):
+                      is_reboot_required_value=None,
+                      commit_required_value=None,
+                      include_commit_required=False):
```
These changes should not be made to build_return_dict.  The commit_required parameter was deprecated and removed from python-dracclient.

mahendraKamble commented:
Hi Chris,

I removed the by_name parameter and added raid_fqdd to get the controller mode.

Thanks for the review comments.

mahendraKamble commented:
Hi Chris,

I modified test cases as per codebase changed.

cdearborn commented on dracclient/resources/uris.py:
```diff
@@ -13,7 +13,7 @@
 """
 Schema definitions and resource URIs for the classes implemented by the DRAC
-WS-Man API.
+WS-Man API.DCIM_RAIDEnumeration
```
Looks like DCIM_RAIDEnumeration was accidentally pasted above.

cdearborn commented on dracclient/tests/test_raid.py:
```diff
@@ -112,6 +113,204 @@ def setUp(self):
+            {'text': test_utils.RAIDEnumerations[
+                uris.DCIM_RAIDEnumeration]['ok']},
+            {'text': test_utils.RAIDEnumerations[
+                uris.DCIM_RAIDString]['ok']}])
```
Don't we need a response for DCIM_RAIDInteger above?

mahendraKamble replied:
Yes, Chris, We don't need a response for DCIM_RAIDInteger. If we add it will not affect so I have added it.

mahendraKamble commented:
Hi Chris,

I have created an upstream patch for it. I will provide its link in a call.

Thanks, Chris

cdearborn commented:
Closing this one since a patch for it has been submitted upstream: https://review.opendev.org/#/c/715172
